fnamchk.1
.\" section 1 man page for fnamchk
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec misfeatures and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH fnamchk 1 "22 October 2022" "fnamchk" "IOCCC tools"
.SH NAME
fnamchk \- IOCCC compressed tarball filename sanity check tool
.SH SYNOPSIS
\fBfnamchk\fP [\-h] [\-v level] [\-q] [\-V] [\-E ext] [\-t|\-u] filepath
.SH DESCRIPTION
\fBfnamchk\fP verifies that an IOCCC compressed tarball is properly named.
.PP
The program validates that the filename is correct, in the form of \fBentry.contest_ID\-entry_num.timestamp.ext\fP where
\fBcontest_ID\fP is either \fItest\fP or a valid UUID (see below), \fBentry_number\fP is an integer from \fB0\fP through \fBMAX_ENTRY_NUM\fP (see \fIlimit_ioccc.h\fP) inclusive, \fBtimestamp\fP is an integer of the number of seconds since the epoch and \fBext\fP is either \fBtxz\fP or the user supplied extension from the \fB\-E\fP option.
.PP
More specifically, the filename \fBmust\fP:
.IP \(bu 4
Start with "\fBentry\fP".
.IP \(bu 4
Followed by "\fB.\fP".
.IP \(bu 4
Followed by either "\fBtest\fP" \fIOR\fP a UUID string in the form of \fBxxxxxxxx\-xxxx\-4xxx\-axxx\-xxxxxxxxxxxx\fP where \fBx\fP is a hexadecimal digit in the range \fB[0\-9a\-f]\fP.
And yes, there is a \fB4\fP (UUID version 4) and an \fBa\fP (UUID variant 1) in there.
.IP \(bu 4
Followed by "\fB\-\fP".
.IP \(bu 4
Followed by a decimal entry number from \fB0\fP through \fBMAX_ENTRY_NUM\fP (see \fIlimit_ioccc.h\fP) inclusive.
.IP \(bu 4
Followed by "\fB.\fP".
.IP \(bu 4
Followed by a positive non\-zero 64\-bit decimal integer.
.IP \(bu 4
Followed by "\fB.\fP".
.IP \(bu 4
Followed by "\fBtxz\fP" (or the user supplied extension from option \fB\-E\fP).
.PP
NOTE: The quotes above should not be in the filename; they're there only to help distinguish the punctuation from the rest of the format.
.PP
It is indirectly invoked via the \fBtxzchk(1)\fP utility which is called by \fBmkiocccentry(1)\fP and it will also be directly executed by the Judges during the judging process of the contests.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-v \fIlevel\fP\fP
Set verbosity level.
Default: 0.
.TP
\fB\-V\fP
Show version and exit.
.TP
\fB\-q\fP
Quiet mode.
Silence msg(), warn(), warnp() if verbosity level is 0.
Default: \fBfalse\fP.
.TP
\fB\-E \fIext\fP\fP
Change extension to extension specified (without the dot!).
This is used for \fBTESTING\fP purposes only!
.TP
\fB\-t\fP
If the filename does not start with the test mode filename format, issue an error.
In other words the filename has to start with \fIentry.test\-\fP or it's an error.
.TP
\fB\-u\fP
If the filename does not start with the normal filename format, issue an error.
In other words if the filename starts with \fIentry.test\-\fP it is an error.
.SH EXIT STATUS
.PP
\fBmain()\fP returns 0 on success; if there's an error the end is not reached.
If there is an error the error message is printed prior to exiting; else the output will be the name of the required directory in the tarball (this is used by \fBtxzchk(1)\fP).
.SH BUGS
.PP
More than 0 humans work on it! :)
.PP
If you have an issue with the tool you can open an issue at
.br
\fI<https://github.com/ioccc\-src/mkiocccentry/issues\>\fP.
.SH EXAMPLES
.PP
For the examples below assume that the \fBMIN_TIMESTAMP\fP is \fB(time_t)1662145368\fP.
.PP
Run the program on the filename \fIentry.test\-0.1562145368.txz\fP which will fail because the timestamp in the filename is less than the minimum timestamp:
.nf
.RS
\fB
 ./fnamchk \-t entry.test\-0.1562145368.txz\fP
.fi
.RE
.PP
Run the program on the filename \fItest.tar\fP, which will fail because the name format is invalid:
.nf
.RS
\fB
 ./fnamchk test.tar\fP
.fi
.RE
.PP
Run the program on the filename \fIentry.test.0.1662145368.txz\fP with the option \fB\-u\fP which will fail because it is a test\-mode filename but the \fB\-u\fP option is specified:
.nf
.RS
\fB
 ./fnamchk \-u entry.test\-0.1662145368.txz\fP
.fi
.RE
.PP
Run the program on the filename \fIentry.test.0.1662145368.txz\fP with the option \fI\-t\fP which will pass because it is a test\-mode file and the timestamp is not less than the minimum:
.nf
.RS
\fB
 ./fnamchk \-t entry.test\-0.1662145368.txz\fP
.fi
.RE
.PP
The above will print "\fBtest\-0\fP".
.PP
Run the program on the file \fIentry.test.0.1662145368.txt\fP with the option \fB\-t\fP, changing the extension to \fItxt\fP via the \fB\-E\fP option, which will pass because test mode is specified with a test filename, the extension matches the \fB\-E\fP option and the timestamp is correct:
.nf
.RS
\fB
 ./fnamchk \-E txt \-t entry.test\-0.1662145368.txt\fP
.fi
.RE
.PP
The above will print "\fBtest\-0\fP".
.PP
Run the program on the file \fIentry.test.0.1662145368.txz\fP with the options \fB\-t\fP, changing the extension to \fItxt\fP via the \fB\-E\fP option which will fail because the extension specified does not match the extension in the filename:
.nf
.RS
\fB
 ./fnamchk \-E txt \-t entry.test\-0.1662145368.txz\fP
.fi
.RE
.PP
It would have passed if the extension was correct because the timestamp in the filename is not less than the minimum timestamp.
.PP
Run the program on the file \fIentry.test.0.1662145368.txz\fP without specifying test\-mode or normal mode which will also pass because the filename format is correct (in this case a test mode filename):
.nf
.RS
\fB
 ./fnamchk entry.test\-0.1662145368.txz\fP
.fi
.RE
.PP
The above will print "\fBtest\-0\fP".


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.

The `chk_foo()` functions in `chk_validate.c` and the `test_foo()` functions in
`entry_util.c` are code complete, although they remain untested and unused.  The
`chkentry` tool, however, is not yet code complete. Later releases of tested
JSON semantic code will no doubt modify these functions.

Improved a number of the ways that JSON field values are checked.  In a number
of cases, code form `mkiocccentry.c` was moved into `test_foo()` functions so
that they could be used by other tools such as the JSON semantic test code.

Added code to generate JSON semantic tables from JSON reference files for
`.info.json` and `.author.json`.  The `jsemcgen.sh` tool manages this by way of
the `jsemtblgen` code generator and header, patch and trailer files (see
`chk.auth.*` and `chk.info.*` files).

Avoided the appearance of attacking any particular individual.  It was not our
intention to disrespect anyone, even though we disagree with some of the
technical decisions.  Where we have fundamental technical disagreements, we
attempted to express those technical disagreements with humour in hopefully a
more fun way.  As also now apologize for how `bison` and `flex` generated code
may look, instead of simply calling it ugly.  As such, we hopefully improved
some of the humour in our code while trying to be nice and friendly to others.

For example now the adjusted dbg levels in JSON parser are:

```
*     At -J 3, only the top level return type and top level tree are printed.
*     At -J 5, intermediate tree return types and tree are printed too.
*     At -J 7, also print grammar progress.
*     At -J 9, also print sorry_text and sorry_lang grammar values.
```

Removed a number of files and added a number of files under the `test_JSON/`
tree.  When the JSON semantic code is being tested in a future release, we
expect more such `test_JSON/` tree changes.

Improved / add a number of man pages.  Updated `README.md`.

Improved and expanded `txzchk`.

Added more test code.  We attempt to detect feathers in tarballs.  :-)

We will neither confirm nor deny the presence of an "Easter egg".
To to so would be "foolish".  :-)

Improved and fixed `vermod.sh` and `reset_tstamp.sh`.  Tested this
code by changing the `MIN_TIMESTAMP` as noted above.  The `MIN_TIMESTAMP`
needed to up updated anyway due to changes in the `.info.json` and
`.author.json` formats.

Made numerous improvements and bug fixes to the Makefile.

Fixed how `picky` is used in by the `make picky` rule for a few special files.

Added multiple rules to the Makefile including but not limited to `make mkref`,
`make legacy_clobber`, and `make legacy_clean` rules.  Applied multiple bug
fixes to the Makefile.

Improved the Makefile to be less impacting for modern systems while trying to
maintain, for as long as we can, compatibility with some older systems.

Attempted to improve compatibility with reasonably modern systems.  We managed
to keep CentOS 7 somewhat supported for now, although we may be forced to drop
support of such an old system before 2024. But if we do reach 2024 we can only
say that given that CentOS 7 will reach EOL in 2024 and CentOS 8 support was
unduly dropped and since < CentOS 7 already was at EOL this should not be a
problem. :-)

Users of very out of date systems can still enter and submit entries to the
IOCCC.  They just might need to find a more modern system to package and submit
their IOCCC entry, however.

Added stub code for `hostchk.sh`.  A future release will include mode tests for
the given hosts.  Future releases will also include a bug report system that
will also use `hostchk.sh`.

Improved the `no-comment` directive in `.info.json` and `.author.json` files.

Improved how `time_t` values are used and printed.  We no longer assume that
`time_t` is signed nor assume it is unsigned: we only assume it's time. :-)

Improved comments in C code about special `seqcexit` comment directives.

Make numerous bug fixes and fixed a fair number of typos.  Nevertheless we claim
this is *NOT* complete. :-)


## Release 0.5 2022-07-10

The `jsemtblgen` tool is code complete.

Added `is_reserved()` function to `rule_count.c`.

Changed the Official iocccsize version from 28.10 2022-03-15 to 28.11
2022-07-10. Updated `test_JSON` tree and `json_teststr.txt` to reflect the new
`IOCCCSIZE_VERSION` value.

Added and improved a number of man pages.

Added `Makefile` rule to build HTML files from all man pages.

Both the `parse_json()` and `parse_json_file()` top level functions of the JSON
parser return a JSON node parse tree when the JSON is valid.

Trimmed the JSON parser to not include code that does not strictly belong to the
generic JSON parser.  Removed dead C code.

Updated dbg facility to version 2.4 2022-06-26.  Version 2.4 2022-06-26 contains
functions to write (form) a message, debug diagnostic message, warning, or error
message into a string buffer:

```c
extern void snmsg(char *str, size_t size, char const *fmt, ...)
extern void sndbg(char *str, size_t size, int level, char const *fmt, ...);
extern void snwarn(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwarnp(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerr(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerrp(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
```


fnamchk.c
/*
 * fnamchk - IOCCC compressed tarball filename sanity check tool
 *
 * "Because even fprintf has a return value worth paying attention to." :-)
 *
 * Copyright (c) 2022 by Landon Curt Noll.  All Rights Reserved.
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright, this permission notice and text
 * this comment, and the disclaimer below appear in all of the following:
 *
 *       supporting documentation
 *       source copies
 *       source works derived from this source
 *       binaries derived from this source or from derived source
 *
 * LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO
 * EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
 * USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 *
 * chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
 *
 * "Because everyone hates when someone gets their name wrong." :-)
 *
 * Share and enjoy! :-)
 */


/* special comments for the seqcexit tool */
/* exit code out of numerical order - ignore in sequencing - ooo */
/* exit code change of order - use new value in sequencing - coo */


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <stdint.h>

/*
 * fnamchk - IOCCC compressed tarball filename sanity check tool
 */
#include "fnamchk.h"


/*
 * definitions
 */
#define REQUIRED_ARGS (1)	/* number of required arguments on the command line */



int
main(int argc, char *argv[])
{
    char const *program = NULL;	/* our name */
    extern char *optarg;	/* option argument */
    extern int optind;		/* argv index of the next arg */
    char *filepath;		/* filepath argument to check */
    char *filename;		/* basename of filepath to check */
    int ret;			/* libc return code */
    char *entry;		/* first '.' separated token - entry */
    char *uuid;			/* first '-' separated token - test or UUID */
    size_t len;			/* UUID length */
    unsigned int a, b, c, d, e, f;	/* parts of the UUID string */
    unsigned int version = 0;	/* UUID version hex character */
    unsigned int variant = 0;	/* UUID variant hex character */
    char guard;			/* scanf guard to catch excess amount of input */
    int entry_num;		/* third .-separated token as a number */
    char *timestamp_str;	/* fourth .-separated token - timestamp */
    intmax_t timestamp;		/* fifth .-separated token as a timestamp */
    char *extension;		/* sixth .-separated token as a filename extension */
    char *ext = "txz";		/* user supplied extension (def: txz): used for testing purposes only */
    int i;
    bool test_mode = false;	/* true ==> force check to test if it's a test entry filename */
    bool uuid_mode = false;	/* true ==> force check to test if it's a UUID entry filename */


    /*
     * parse args
     */
    program = argv[0];
    while ((i = getopt(argc, argv, "hv:qVtquE:")) != -1) {
	switch (i) {
	case 'h':		/* -h - print help to stderr and exit 0 */
	    usage(1, "-h help mode", program); /*ooo*/
	    not_reached();
	    break;
	case 'v':		/* -v verbosity */
	    /*
	     * parse verbosity
	     */
	    verbosity_level = parse_verbosity(program, optarg);
	    break;
	case 'q':
	    msg_warn_silent = true;
	    break;
	case 'V':		/* -V - print version and exit */
	    print("%s\n", FNAMCHK_VERSION);
	    exit(0); /*ooo*/
	    not_reached();
	    break;
	case 't': /* force check to verify that the filename is a test entry filename */
	    test_mode = true;
	    break;
	case 'u': /* force check to verify that the filename is a UUID (real) entry filename */
	    uuid_mode = true;
	    break;
	case 'E': /* force extension check to be optarg instead of "txz": used for test suite for txzchk */
	    ext = optarg;
	    break;
	default:
	    usage(1, "invalid -flag", program); /*ooo*/
	    not_reached();
	    break;
	 }
    }
    /* must have the exact required number of args */
    if (argc - optind != REQUIRED_ARGS) {
	usage(1, "wrong number of arguments", program); /*ooo*/
	not_reached();
    }

    /* -t and -u cannot be used together as the tests they enable conflict */
    if (test_mode && uuid_mode) {
	err(1, __func__, "-t and -u cannot be used together");
	not_reached();
    }

    /* collect arg */
    filepath = argv[optind];
    dbg(DBG_LOW, "filepath: %s", filepath);

    /*
     * obtain the basename of the path to examine
     */
    filename = base_name(filepath);
    dbg(DBG_LOW, "filename: %s", filename);

    /*
     * first .-separated token must be entry
     */
    entry = strtok(filename, ".");
    if (entry == NULL) {
	err(2, __func__, "first strtok() returned NULL");
	not_reached();
    }
    if (strcmp(entry, "entry") != 0) {
	err(3, __func__, "filename does not start with \"entry.\": %s", filepath);
	not_reached();
    }
    dbg(DBG_LOW, "filename starts with \"entry.\": %s", filename);

    /*
     * second '.' separated token must be test or a UUID
     */
    uuid = strtok(NULL, ".");
    if (uuid == NULL) {
	err(4, __func__, "nothing found after first '.'  separated token of entry");
	not_reached();
    }
    len = strlen(uuid);

    /*
     * parse a test-entry_num IOCCC contest ID
     */
    if (strncmp(uuid, "test-", LITLEN("test-")) == 0) {
	/* if it starts as "test-" and -u was specified it's an error */
	if (uuid_mode) {
	    err(5, __func__, "-u specified and entry starts as a test mode filename");
	    not_reached();
	}

	/*
	 * NOTE: we prevent seqcexit from modifying the exit code because the
	 * txzchk_test.sh script has a test file where it expects this error.
	 */
	if (len != LITLEN("test-")+MAX_ENTRY_CHARS) {
	    err(6, __func__, "second '-' separated token length: %ju != %ju: %s",
			     (uintmax_t)len, (uintmax_t)(LITLEN("test-")+MAX_ENTRY_CHARS), filepath); /*ooo*/
	    not_reached();
	}
	ret = sscanf(uuid, "test-%d%c", &entry_num, &guard);
	if (ret != 1) {
	    err(7, __func__, "second '.' separated non-test not in test-entry_number format: %s", filepath);
	    not_reached();
	}
	dbg(DBG_LOW, "entry ID is test: %s", uuid);
	if (entry_num < 0) {
	    err(8, __func__, "third '.' separated entry number: %d is < 0: %s", entry_num, filepath);
	    not_reached();
	}
	if (entry_num > MAX_ENTRY_NUM) {
	    err(9, __func__, "second '-' separated entry number: %d is > %d: %s", entry_num, MAX_ENTRY_NUM, filepath);
	    not_reached();
	}
	dbg(DBG_LOW, "entry number is valid: %d", entry_num);

    /*
     * parse a UUID-entry_num IOCCC contest ID
     */
    } else {
	/*
	 * if -t is specified and we get here (filename does not start with
	 * "entry.test-" then it's an error.
	 */
	if (test_mode) {
	    err(10, __func__, "-t specified and entry does not start with \"entry.test-\"");
	    not_reached();
	}

	/*
	 * NOTE: we prevent seqcexit from modifying the exit code because the
	 * txzchk_test.sh script has a test file where it expects this error.
	 */
	if (len != UUID_LEN+1+MAX_ENTRY_CHARS) {
	    err(11, __func__, "second '-' separated token length: %ju != %ju: %s",
			     (uintmax_t)len, (uintmax_t)(UUID_LEN+1+MAX_ENTRY_CHARS), filepath); /*ooo*/
	    not_reached();
	}
	ret = sscanf(uuid, "%8x-%4x-%1x%3x-%1x%3x-%8x%4x-%d%c", &a, &b, &version, &c, &variant, &d, &e, &f, &entry_num, &guard);
	if (ret != 9) {
	    err(12, __func__, "second '.' separated non-test not in UUID-entry_number format: %s", filepath);
	    not_reached();
	}
	if (version != UUID_VERSION) {
	    err(13, __func__, "second '.' separated UUID token version %x != %x: %s", version, UUID_VERSION, filepath);
	    not_reached();
	}
	if (variant != UUID_VARIANT) {
	    err(14, __func__, "second '.' separated UUID token variant %x != %x: %s", variant, UUID_VARIANT, filepath);
	    not_reached();
	}
	dbg(DBG_LOW, "entry ID is a valid UUID: %s", uuid);
	if (entry_num < 0) {
	    err(15, __func__, "third '.' separated entry number: %d is < 0: %s", entry_num, filepath);
	    not_reached();
	}
	if (entry_num > MAX_ENTRY_NUM) {
	    err(16, __func__, "second '-' separated entry number: %d is > %d: %s", entry_num, MAX_ENTRY_NUM, filepath);
	    not_reached();
	}
	dbg(DBG_LOW, "entry number is valid: %d", entry_num);
    }

    /*
     * third '.' separated token must be a valid timestamp
     */
    timestamp_str = strtok(NULL, ".");
    if (timestamp_str == NULL) {
	err(17, __func__, "nothing found after second '.' separated token of entry number");
	not_reached();
    }
    ret = sscanf(timestamp_str, "%jd%c", &timestamp, &guard);
    if (ret != 1) {
	err(18, __func__, "third '.' separated token: %s is not a timestamp: %s", timestamp_str, filepath);
	not_reached();
    }
    if (timestamp < MIN_TIMESTAMP) {
	err(19, __func__, "third '.' separated timestamp: %jd is < %jd: %s", timestamp, (intmax_t)MIN_TIMESTAMP, filepath);
	not_reached();
    }
    dbg(DBG_LOW, "timestamp is valid: %jd", timestamp);

    /*
     * fourth .-separated token must be the filename extension
     */
    extension = strtok(NULL, ".");
    if (extension == NULL) {
	err(20, __func__, "nothing found after third '.' separated token of timestamp");
	not_reached();
    }
    if (strcmp(extension, ext) != 0) {
	err(21, __func__, "final fourth '.' separated token filename extension: %s != %s: %s", extension, ext, filepath);
	not_reached();
    }
    dbg(DBG_LOW, "filename extension is valid: %s", extension);

    /*
     * filepath must use only POSIX portable filename and + chars /
     */
    if (posix_plus_safe(filepath, false, true, false) == false) {
	err(22, __func__, "filepath: posix_plus_safe(%s, false, true, false) is false", filepath);
	not_reached();
    }

    /*
     * filename must use only lower case POSIX portable filename and + chars
     */
    if (posix_plus_safe(filename, true, false, true) == false) {
	err(23, __func__, "basename: posix_plus_safe(%s, true, false, true) is false", filename);
	not_reached();
    }

    /*
     * All is OK with the filepath - print entry directory basename
     */
    dbg(DBG_LOW, "filepath passes all checks: %s", filepath);
    errno = 0;		/* pre-clear errno for errp() */
    ret = printf("%s\n", uuid);
    if (ret <= 0) {
	errp(24, __func__, "printf of entry directory basename failed");
	not_reached();
    }

    /*
     * free filename allocated from base_name()
     */
    if (filename != NULL) {
	free(filename);
	filename = NULL;
    }

    /*
     * All Done!!! - Jessica Noll, age 2
     *
     */
    exit(0); /*ooo*/
}


/*
 * usage - print usage to stderr
 *
 * Example:
 *      usage(3, "missing required argument(s), program: %s", program);
 *
 * given:
 *	exitcode        value to exit with
 *	str		top level usage message
 *	program		our program name
 *
 * NOTE: We warn with extra newlines to help internal fault messages stand out.
 *       Normally one should NOT include newlines in warn messages.
 *
 * This function does not return.
 */
static void
usage(int exitcode, char const *str, char const *prog)
{
    /*
     * firewall
     */
    if (str == NULL) {
	str = "((NULL str))";
	warn(__func__, "\nin usage(): program was NULL, forcing it to be: %s\n", str);
    }
    if (prog == NULL) {
	prog = "((NULL prog))";
	warn(__func__, "\nin usage(): program was NULL, forcing it to be: %s\n", prog);
    }

    /*
     * print the formatted usage stream
     */
    fprintf_usage(DO_NOT_EXIT, stderr, "%s\n", str);
    fprintf_usage(exitcode, stderr, usage_msg, prog, DBG_DEFAULT, FNAMCHK_VERSION);
    exit(exitcode); /*ooo*/
    not_reached();
}
